src/protocols/secure_channel/PairingSession.cpp

```diff
@@ -58,6 +59,24 @@ void PairingSession::Finish()
 {
     Transport::PeerAddress address = mExchangeCtxt->GetSessionHandle()->AsUnauthenticatedSession()->GetPeerAddress();
 
+#if INET_CONFIG_ENABLE_TCP_ENDPOINT
+    // Release the ExchangeContext that was retained at the beginning of CASE
+    // initiation(both at initiator and responder side). The Retain() was
+    // necessary to prevent the automatic release of the ExchangeContext
+    // after sending the last message of the Exchange and before the call
+    // to Finish()
+    if (address.GetTransportType() == Transport::Type::kTcp)
+    {
+        // Fetch the connection for the unauthenticated session used to setup
+        // the secure session.
+        Transport::ActiveConnectionState * conn = mExchangeCtxt->GetSessionHandle()->AsUnauthenticatedSession()->GetTCPConnection();
+
+        // Associate the connection with the secure session being activated.
+        mSecureSessionHolder->AsSecureSession()->SetTCPConnection(conn);
+
+        mExchangeCtxt->Release();
```

Comment:
What guarantees that we reached one of the Retain() calls before now?  I suspect nothing.

Reply:
Changed to ExchangeHandle.